src/Microsoft.ML.TimeSeries/RootCauseAnalyzer.cs

```diff
@@ -0,0 +1,764 @@
+        {
+            Dictionary<string, Object> newDim = new Dictionary<string, Object>(dimension);
+            newDim[key] = value;
+            return newDim;
+        }
+
+        private bool StopAnomalyComparison(double preTotal, double parent, double current, double pre)
+        {
+            if (Math.Abs(preTotal) < Math.Abs(parent) * _anomalyDeltaThreshold)
+            {
+                return false;
+            }
+
+            return Math.Abs(pre) / Math.Abs(current) > _anomalyPreDeltaThreshold;
+        }
+
+        private bool ShouldSeperateAnomaly(double total, double parent, int totalSize, int size)
+        {
+            if (Math.Abs(total) < Math.Abs(parent) * _anomalyDeltaThreshold)
+            {
+                return false;
+            }
+
+            if (size == totalSize && size == 1)
+            {
+                return true;
+            }
+
+            return size <= totalSize * _anomalyRatioThreshold;
+        }
+
+        private double GetDimensionEntropy(Dictionary<string, int> pointDis, Dictionary<string, int> anomalyDis)
+        {
+            int total = GetTotalNumber(pointDis);
+            double entropy = 0;
+
+            foreach (string key in anomalyDis.Keys)
+            {
+                double dimEntropy = GetEntropy(pointDis[key], anomalyDis[key]);
+                entropy += dimEntropy * pointDis[key] / total;
+            }
+
+            return entropy;
+        }
+
+        private double GetDimensionInstrinsicValue(Dictionary<string, int> pointDis)
+        {
+            int total = GetTotalNumber(pointDis);
+            double instrinsicValue = 0;
+
+            foreach (string key in pointDis.Keys)
+            {
+                instrinsicValue -= Log2((double)pointDis[key] / total) * (double)pointDis[key] / total;
+            }
+
+            return instrinsicValue;
+        }
+
+        private int GetTotalNumber(Dictionary<string, int> distribution)
+        {
+            int total = 0;
+            foreach (int num in distribution.Values)
+            {
+                total += num;
+            }
+            return total;
+        }
+
+        private void UpdateDistribution(Dictionary<string, int> distribution, List<Point> points, string dimKey)
+        {
+            foreach (Point point in points)
+            {
+                string dimVal = (string)point.Dimension[dimKey];
+                if (!distribution.ContainsKey(dimVal))
+                {
+                    distribution.Add(dimVal, 0);
+                }
+                distribution[dimVal] = distribution[dimVal] + 1;
+            }
+        }
+
+        private double Log2(double val)
+        {
+            if (Double.IsNaN(val))
+            {
+                return 0;
+            }
+
+            return Math.Log(val) / Math.Log(2);
+        }
+
+        private static bool ContainsAll(Dictionary<string, Object> bigDic, Dictionary<string, Object> smallDic)
+        {
```

Comment:
suggestion: Please consider changing the names of the variables to bigDictionary and smallDictionary

Reply:
> suggestion: Please consider changing the names of the variables to bigDictionary and smallDictionary

updated